examples/vector_database/scripts/compute_vectors.py

```diff
@@ -0,0 +1,240 @@
+
+        # Convert to numpy arrays
+        embeddings = features.cpu().numpy()
+
+        # Convert original images to base64
+        images_base64 = {}
+        for idx, img in zip(indices, original_images):
+            buffered = BytesIO()
+            img.save(buffered, format="JPEG")
+            img_str = base64.b64encode(buffered.getvalue()).decode()
+            images_base64[idx] = img_str
+
+        # Return both embeddings and images
+        return [(idx, pickle.dumps((images_base64[idx], arr)))
+                for idx, arr in zip(indices, embeddings)]
+
+    async def run(self):
+        """
+        Run the batch processing pipeline without appending to existing files.
+        Each partition is written atomically using a temporary file.
+        """
+        results = []
+        partition_counter = 0
+
+        async for idx, input_data in self.do_data_loading():
+            self._current_batch.append((idx, input_data))
+            if len(self._current_batch) >= self.batch_size:
+                batch_results = await self.do_batch_processing(
+                    self._current_batch)
+                results.extend(batch_results)
+                self._current_batch = []
+
+                if len(results) >= self.checkpoint_size:
+                    # Convert results to a DataFrame
+                    df = pd.DataFrame(results, columns=["idx", "output"])
+
+                    # Create temporary and final file paths
+                    final_path = f"{self.output_path}_part_{partition_counter}.parquet"
+                    temp_path = f"/tmp/{partition_counter}.tmp"
+
+                    # Write to temporary file first
+                    df.to_parquet(temp_path, engine="pyarrow", index=False)
+
+                    # Copy from temp to final destination
+                    shutil.copy2(temp_path, final_path)
+                    os.remove(temp_path)  # Clean up temp file
+
+                    logging.info(
+                        f"Saved partition {partition_counter} to {final_path} with {len(df)} rows"
+                    )
+                    results.clear()
+                    partition_counter += 1
+
+        # Process any remaining items in the batch
+        if self._current_batch:
+            batch_results = await self.do_batch_processing(self._current_batch)
+            results.extend(batch_results)
+
+        # Write the final partition if there are any leftover results
+        if results:
+            df = pd.DataFrame(results, columns=["idx", "output"])
+            final_path = f"{self.output_path}_part_{partition_counter}.parquet"
+            temp_path = f"/tmp/{partition_counter}.tmp"
+
+            df.to_parquet(temp_path, engine="pyarrow", index=False)
+            shutil.copy2(temp_path, final_path)
+            os.remove(temp_path)  # Clean up temp file
+
+            logging.info(
+                f"Saved final partition {partition_counter} to {final_path}")
```

Comment:
since this is the same as above, maybe we can break this out to a method as well.